### PR TITLE
Add kubestash controller for changing kubeDB phase

### PR DESCRIPTION
**Pull request by anisurrahman75**
Signed-off-by: Anisur Rahman <anisur@appscode.com>

### CODE REVIEW

File: pkg/controller/initializer/restore/util.go [hunk 1]
```diff
@@ -281,27 +295,57 @@ func (c *Controller) waitUntilStashInstalled(stopCh <-chan struct{}) error {
...
 	}
+
 	var owner *metav1.OwnerReference
-	if matched, err := targetOfGroupKind(ri.target.Ref, appcat.GroupName, ab.ResourceKindApp); err == nil && matched {
-		appBinding, err := c.AppCatalogClient.AppcatalogV1alpha1().AppBindings(ri.do.Namespace).Get(context.TODO(), ri.target.Ref.Name, metav1.GetOptions{})
-		if err != nil {
-			return err
+	if ri.stashTarget != nil {
+		if matched, err := targetOfGroupKind(ri.stashTarget.Ref, appcat.GroupName, ab.ResourceKindApp); err == nil && matched || ri.kubeStashTarget != nil {
+			appBinding, err := c.AppCatalogClient.AppcatalogV1alpha1().AppBindings(ri.do.Namespace).Get(context.TODO(), ri.stashTarget.Ref.Name, metav1.GetOptions{})
```
Comment (hmsayem) [inline pkg/controller/initializer/restore/util.go:329]: remove `ri.kubeStashTarget != nil` if not necessary

File: pkg/controller/initializer/restore/util.go [hunk 2]
```diff
@@ -281,27 +295,57 @@ func (c *Controller) waitUntilStashInstalled(stopCh <-chan struct{}) error {
...
+			if err != nil {
+				return err
+			}
+			owner = metav1.GetControllerOf(appBinding)
+		} else if matched, err := targetOfGroupKind(ri.stashTarget.Ref, apps.GroupName, sapis.KindStatefulSet); err == nil && matched {
+			sts, err := c.AppCatalogClient.AppcatalogV1alpha1().AppBindings(ri.do.Namespace).Get(context.TODO(), ri.stashTarget.Ref.Name, metav1.GetOptions{})
+			if err != nil {
```
Comment (hmsayem) [inline pkg/controller/initializer/restore/util.go:336]: `ri.do.Namespace` should be replaced with `ri.kubeStashTarget.Namespace`

**Review by hmsayem — CHANGES_REQUESTED** [review]

**Review by ishtiaqhimel — DISMISSED** [review]

File: pkg/controller/initializer/restore/util.go [hunk 3]
```diff
@@ -281,27 +295,57 @@ func (c *Controller) waitUntilStashInstalled(stopCh <-chan struct{}) error {
...
+	return wait.PollImmediateUntil(time.Second*10, func() (bool, error) {
+		return discovery.ExistsGroupKinds(c.Client.Discovery(),
+			schema.GroupKind{Group: storageapi.GroupVersion.Group, Kind: storageapi.ResourceKindBackupStorage},
+			schema.GroupKind{Group: storageapi.GroupVersion.Group, Kind: storageapi.ResourceKindRepository},
+			schema.GroupKind{Group: storageapi.GroupVersion.Group, Kind: storageapi.ResourceKindSnapshot},
+			schema.GroupKind{Group: coreapi.GroupVersion.Group, Kind: coreapi.ResourceKindBackupBatch},
+			schema.GroupKind{Group: coreapi.GroupVersion.Group, Kind: coreapi.ResourceKindBackupBlueprint},
+			schema.GroupKind{Group: coreapi.GroupVersion.Group, Kind: coreapi.ResourceKindBackupConfiguration},
+			schema.GroupKind{Group: coreapi.GroupVersion.Group, Kind: coreapi.ResourceKindBackupSession},
+			schema.GroupKind{Group: coreapi.GroupVersion.Group, Kind: coreapi.ResourceKindRestoreSession},
```
Comment (hmsayem) [inline pkg/controller/initializer/restore/util.go:311]: Add RetentionPolicy

**Review by hmsayem — CHANGES_REQUESTED** [review]

File: pkg/controller/initializer/restore/controller.go [hunk 4]
```diff
@@ -26,37 +27,45 @@ import (
...
-	invoker    core.TypedLocalObjectReference
-	target     *v1beta1.RestoreTarget
-	phase      v1beta1.RestorePhase
-	do         dmcond.DynamicOptions
-	invokerUID types.UID
+	invoker         core.TypedLocalObjectReference
+	stashTarget     *v1beta1.RestoreTarget
+	kubeStashTarget *kmapi.TypedObjectReference
+	stashPhase      v1beta1.RestorePhase
+	kubeStashPhase  coreapi.RestorePhase
```
Comment (ArnobKumarSaha) [inline pkg/controller/initializer/restore/controller.go:66]: We should make this struct more organized . like 
```
type restoreInfo struct {
	invoker     core.TypedLocalObjectReference
	stash       *stashInfo
	kubestash   *kubestashInfo
	do          dmcond.DynamicOptions
	invokerUID  types.UID
}

type stashInfo struct {
	target *v1beta1.RestoreTarget
	phase v1beta1.RestorePhase
}

type kubestashInfo struct {
	target *kmapi.TypedObjectReference
	phase coreapi.RestorePhase
}
```

**Review by ArnobKumarSaha — CHANGES_REQUESTED** [review]

**Review by ArnobKumarSaha — APPROVED** [review]

**Pull request merged**